### PR TITLE
Perf: keep DeepSeek MTP tensors device-resident

### PR DESCRIPTION
**Pull request by Little-oil**
## Summary

This PR optimizes two independent sources of MTP dispatch overhead:

1. **Output residency:** keep `hidden_out`, `pre_hc_hidden_out`, and `logits` on device, eliminating their per-dispatch D2H copy-back and the D1 H2D upload of `hidden_out`.
2. **Static-weight residency:** keep the rank-sharded MTP weights/constants on their consuming cards, eliminating repeated H2D weight staging and the resulting rank launch skew.

It also:

- allocates pure `Out` resident buffers without uploading host zero initialization
- adds explicit `validate` and `device-resident` output modes
- reads resident outputs back once when correctness validation is requested
- extends runner/spec tests for resident output allocation and readback

## Measurement setup

Unless noted otherwise:

- devices: 0,2 (rank0/dev0 and rank1/dev2)
- seed: 0
- L2 swimlane: disabled
- same compiled runtime artifact
- times below are per dispatch
- two-rank aggregate values in section 1 are arithmetic means

Resident tensors are uploaded once per `prepare()` lifecycle and reused across validation, warmup, and timed rounds. Therefore the results below measure steady-state dispatch improvement; they do not claim to remove first model-load time.

### What changed

- `hidden_out` is shared directly from D0 to D1 on device.
  - removes 1 MiB/rank D0 D2H
  - removes 1 MiB/rank D1 H2D
- `pre_hc_hidden_out` remains on device.
  - removes 8 MiB/rank D0 D2H
- `logits` remains on device.
  - removes 4,136,960 B/rank D1 D2H
- pure output resident buffers use uninitialized device allocation, avoiding a one-time 12.945 MiB/rank host-zero upload

### Before/after timing

| Metric, two-rank average | Before | After | Improvement |
|---|---:|---:|---:|
| D0 `validate` | 2,309.3 us | 9.6 us | -99.6% |
| D0 `copy_back` | 2,301.3 us | 1.2 us | -99.9% |
| D1 `bind.args` | 115.7 us | 22.9 us | -80.2% |
| D1 `validate` | 1,183.3 us | 7.5 us | -99.4% |
| D1 `copy_back` | 1,176.5 us | 0.09 us | ~-100% |
| D0 + D1 `validate` | 3,492.6 us | 17.1 us | -99.5% |
| D1 `simpler_run` | 3,131.4 us | 1,534.2 us | -51.0% |

The child-runtime `validate` spans above no longer contain per-dispatch output D2H. In `--output-mode validate`, the harness still performs one explicit final readback before golden comparison. In `--output-mode device-resident`, that final readback is skipped as well.

### What changed

47 D0 static weights/constants are marked `resident="stacked"`; `lm_head_weight` remains resident for D1. Shard `r` is uploaded once to logical worker `r` and reused by all dispatches in the prepared runtime.

Only these seven dynamic request inputs remain on the per-dispatch H2D path:

- `hidden_states`
- `prev_hidden_states`
- `ori_block_table`
- `ori_slot_mapping`
- `position_ids`
- `input_ids`
- `logit_row_indices`

### Cold-dispatch staging and rank skew

| Metric | Weights not resident | Routed W1/W3/W2 resident | All static weights resident |
|---|---:|---:|---:|
| D0 H2D per rank | 985.505 MiB | 217.505 MiB | **9.003 MiB** |
| rank0 D0 `bind.args` | 173.774 ms | 51.268 ms | **2.845 ms** |
| rank1 D0 `bind.args` | 205.795 ms | 52.480 ms | **2.797 ms** |
| two-rank average `bind.args` | 189.784 ms | 51.874 ms | **2.821 ms** |
| rank-to-rank D0 runner-entry skew | 33.270 ms | 1.946 ms | **0.487 ms** |
| rank0 D0 `simpler_run` | 266.546 ms | 60.258 ms | **9.976 ms** |
| rank1 D0 `simpler_run` | 265.911 ms | 59.538 ms | **9.437 ms** |

Final improvement versus weights not resident:

- repeated H2D bytes: 985.505 MiB -> 9.003 MiB (**-99.09%**)
- average `bind.args`: 189.784 ms -> 2.821 ms (**-98.51%**)
- D0 runner-entry skew: 33.270 ms -> 0.487 ms (**-98.54%**)
- rank0/rank1 D0 `simpler_run`: 266.546/265.911 ms -> 9.976/9.437 ms

The original ~33 ms rank skew was therefore caused by unequal completion time while both ranks synchronously staged almost 1 GiB of inputs, not by validation, swimlane collection, or extra kernel computation.

### Steady-state benchmark

`PYPTO_BENCH_WARMUP=2`, `PYPTO_BENCH_ROUNDS=5`, `--output-mode device-resident`:

| Metric | Weights not resident | Routed W1/W3/W2 resident | All static weights resident | Final improvement |
|---|---:|---:|---:|---:|
| `host_union_mean` | 305.120 ms | 59.842 ms | **8.114 ms** | **-97.34%** |
| `host_mean` | 303.618 ms | 59.710 ms | **8.031 ms** | **-97.35%** |
| `effective` median | 30.312 ms | 5.901 ms | **3.867 ms** | **-87.24%** |

A diagnostic run confirmed the final dispatch copy set:

- D0: 6 H2D copies/rank, 9,440,256 B
- D1: 1 H2D copy/rank, 32 B
- total: 9,440,288 B = 9.003 MiB/rank
- no static weight appears on the dispatch H2D path

## Validation

- `pytest -q tests/golden` — 170 passed
- `ruff check golden/runner.py golden/spec.py models/deepseek_v4_flash_mtp/prefill_mtp.py tests/golden/test_runner.py` — passed
- output-resident correctness, devices 0,2 — all four outputs passed (`task_20260804_012422_40358377296`)
- routed-weight resident correctness — all four outputs passed (`task_20260804_023644_383849919652`)
- all-static resident correctness — all four outputs passed (`task_20260804_024241_7180818592`)
- routed-weight benchmark — passed (`task_20260804_023833_40809802404`)
- all-static benchmark — passed (`task_20260804_024431_34203224591`)
- final H2D diagnostic run — passed (`task_20260804_024554_40443710232`)

### CODE REVIEW

File: .github/actions/pypto-serving-tests/action.yml [hunk 1]
```diff
@@ -65,7 +65,8 @@ runs:
       uses: actions/checkout@v4
       with:
         repository: hw-native-sys/pypto-serving
-        ref: main
+        # pypto-serving#136 follows pypto-lib's flattened model directories.
+        ref: 2e6fbd950dc3eb656a9789b00701ed57324d90f8
```
Comment (zhangqi-chen) [inline .github/actions/pypto-serving-tests/action.yml]: serving has been fixed, remove this

File: golden/runner.py [hunk 2]
```diff
@@ -763,10 +763,57 @@ def _share_in_place(tensors: dict[str, torch.Tensor]) -> None:
         tensors[name] = t.cpu().contiguous().share_memory_()
 
 
+def _strip_ssa_suffix(name: str) -> str:
```
Comment (zhangqi-chen) [inline golden/runner.py:766]: try keep runner unchanged.

File: models/deepseek_v4_flash_mtp/prefill_mtp.py [hunk 3]
```diff
@@ -691,13 +743,27 @@ def main():
...
+    )
     parser.add_argument("--enable-l2-swimlane", type=int, nargs="?", const=1, default=0, choices=(0, 1, 2))
     parser.add_argument("--enable-scope-stats", action="store_true", default=False)
     parser.add_argument("--compile-only", action="store_true", default=False)
     parser.add_argument("--dump-passes", action="store_true", default=False)
     parser.add_argument("--runtime-dir", type=str, default=None)
     args = parser.parse_args()
 
+    torch.manual_seed(args.seed)
+    print(f"[RUN] seed={args.seed}", flush=True)
```
Comment (zhangqi-chen) [inline models/deepseek_v4_flash_mtp/prefill_mtp.py]: do not use seed

File: .github/actions/pypto-serving-tests/action.yml (hunk 1 above)
Comment (Little-oil) [reply .github/actions/pypto-serving-tests/action.yml]: Fixed in 5f3e6e5 — restored the pypto-serving checkout to ref: main and removed the obsolete revision pin.

File: models/deepseek_v4_flash_mtp/prefill_mtp.py (hunk 3 above)
Comment (Little-oil) [reply models/deepseek_v4_flash_mtp/prefill_mtp.py]: Fixed in 5f3e6e5 — removed the --seed option, torch.manual_seed(), and the seed log output.

File: golden/runner.py (hunk 2 above)
Comment (Little-oil) [reply golden/runner.py:766]: 这部分优化发生在 compiled.prepare() 之后、第一次 dispatch 之前的 resident 输出 buffer 初始化：hidden_out、pre_hc_hidden_out、logits 被编译为 pure Out 时使用 init=None 分配，跳过两卡合计 25.89 MiB 的 host 零值 H2D。0/2 卡、同一 artifact 做了两次独立 ABBA/BAAB A/B，old alloc + zero H2D 对比 new alloc only 的 paired median 分别节省 7,557 us 和 10,508 us / prepare。这个收益每个 prepare 发生一次，稳态每次 dispatch 是 0 us。
